Feature/seance reservation

MonikaLukowska opened the pull request:
This PR add ticket reservation feature.  Implementation includes:

- Reservation model and migration
- Ticket model and migration
- CreateReservation service object
- Reservation controller
- Custom validator for seats availability
- Model tests


https://user-images.githubusercontent.com/56922072/191772025-14042e9f-68e4-49f0-9cb2-585af4b8e597.mov

MonikaLukowska committed:
feat: add Reservation model

MonikaLukowska committed:
feat: add Ticket model with custom validator

This commit also adds associations to seance,adds method mapping seats to hash, where value keeps informationa about availability  and method creating all seats array

MonikaLukowska committed:
feat:add routes for reservations

MonikaLukowska committed:
feat: add CreateReservation service

MonikaLukowska committed:
feat: add reservations controller with new and create methods

MonikaLukowska committed:
feat:add views for tickets reservation

MonikaLukowska committed:
test:add model tests for reservation and ticket

MonikaLukowska committed:
fix:seat availability proper expectation format

kodowicz commented on app/controllers/reservations_controller.rb:
```diff
@@ -0,0 +1,21 @@
+    else
+      redirect_back fallback_location: new_seance_reservation_path(seance),
+                    alert: ticket_reservation.errors.join
+    end
```
```suggestion
    reservation = CreateReservation.new(params:).call
    if reservation
      redirect_to seances_path, notice: t('.notice')
    else
      redirect_back fallback_location: new_seance_reservation_path(seance),
                    alert: reservation.errors.join
    end
```

kodowicz commented on app/models/reservation.rb:
```diff
@@ -0,0 +1,6 @@
+  belongs_to :seance
+  has_many :tickets, dependent: :destroy
+
+  validates :email, presence: true
```
if you don't have devise right now you could at least validate email format

MonikaLukowska replied:
good point, I will add it.

kodowicz commented on db/migrate/20220921093216_create_tickets.rb:
```diff
@@ -0,0 +1,10 @@
+class CreateTickets < ActiveRecord::Migration[7.0]
+  def change
+    create_table :tickets do |t|
```
where do you have a price? 🤔

MonikaLukowska replied:
We agreed at DB modeling workshops that it can be a part of seance

kodowicz commented on app/models/seance.rb:
```diff
@@ -1,11 +1,22 @@
+  end
+
+  def taken_seats
+    tickets.map(&:seat)
```
```suggestion
    tickets.pluck(:seat)
```

kodowicz commented on app/models/seance.rb:
```diff
@@ -1,11 +1,22 @@
   scope :filtered_by_day, ->(date) { where('Date(start_time) = ?', date) }
+
+  def seats_hash
+    hall.all_seats.index_with do |seat|
+      taken_seats.include?(seat) ? 'reserved' : 'available'
+    end
```
I think it would be more readable if you create an array with hashes like:
```ruby
[{ seat: 10, status: "available" }]
 ```
instead of 
```
{ 10: "available" }
```

MonikaLukowska committed:
feat:add reservation email validation

MonikaLukowska committed:
test:add reservation email validation tests

MonikaLukowska committed:
ref:change variable name to more accurate

wijanko commented on app/controllers/reservations_controller.rb:
```diff
@@ -0,0 +1,21 @@
+  end
+
+  def create
+    reservation = CreateReservation.new(params:)
```
You are passing `params` as named parameter, I was not sure about this syntax, but you attach the video and proved that it work. 🚀 
```suggestion
    reservation = CreateReservation.new(params: params)
```

You can consider passing explicit arguments to `CreateReservation`. Benefits I see in this approach:
- you can see from controller what exact arguments reservation needs to be created
- you can reuse the `CreateReservation` service in some different use case (not everywhere you have access to this specific `params`).
In situation when creating reservation needs a lot of arguments you can pass it as one `params` (or any other compact struct).
This is just my suggestion, your solution is also correct, you can chose what you want.

```suggestion
    reservation = CreateReservation.new(email: email, seance_id: seance_id seats: seats)
```

wijanko commented on app/validators/seat_availability_validator.rb:
```diff
@@ -0,0 +1,9 @@
+class SeatAvailabilityValidator < ActiveModel::Validator
+  def validate(record)
+    status = record.seance.seats_hash[record.seat]
```
Consider using better naming. `seats_hash` does not tell much about real purpose of this method.
Maybe
```suggestion
    status = record.seance.seats_availability[record.seat]
```
WDYT?

MonikaLukowska replied:
I agree, more descriptive:)

wijanko commented on app/validators/seat_availability_validator.rb:
```diff
@@ -0,0 +1,9 @@
+class SeatAvailabilityValidator < ActiveModel::Validator
+  def validate(record)
+    status = record.seance.seats_hash[record.seat]
+    seat = record.seat
```
It is good practice to declare variables before uses if possible. If status will be `reserved` this `seat`  variable wont do anything.
You can move it behind return or use `record.seat `directly in error.message

wijanko reviewed (approved):
Looks good to me. Please consider my comments.
Great that you attached video.
🥁

MonikaLukowska committed:
ref:change seats_hash method to return array of hashes

Instead returning hash of avaialabilities it will return an array of hashes where keys are seat and status. Method name was alse changed to more descriptive

MonikaLukowska committed:
chore: disable short hash syntax enforcment

MonikaLukowska committed:
fix:adapt method name in reservation form

MonikaLukowska committed:
ref:initialize service with explicit parameters

MonikaLukowska committed:
ref:adjust validator parameter name
